src/react/useMembers.ts

```diff
@@ -0,0 +1,121 @@
+import { useEffect, useRef, useState } from 'react';
+import { useSpace } from './useSpace.js';
+import { isArray, isFunction, isString } from '../utilities/is.js';
+
+import type { ErrorInfo } from 'ably';
+import type { Space, SpaceMember } from '..';
+import type { UseSpaceOptions } from './types.js';
+
+interface UseMembersResult {
+  space?: Space;
+  /**
+   * All members present in the space
+   */
+  members: SpaceMember[];
+  /**
+   * All members present in the space excluding the member associated with the spaces client
+   */
+  others: SpaceMember[];
+  /**
+   * The member associated with the spaces client
+   */
+  self: SpaceMember | null;
+  channelError: ErrorInfo | null;
+  connectionError: ErrorInfo | null;
+}
+
+type UseMembersCallback = (params: SpaceMember) => void;
+
+type MembersEvent = 'leave' | 'enter' | 'update' | 'updateProfile' | 'remove';
+
+function useMembers(callback?: UseMembersCallback, options?: UseSpaceOptions): UseMembersResult;
+function useMembers(
+  event: MembersEvent | MembersEvent[],
+  callback: UseMembersCallback,
+  options?: UseSpaceOptions,
+): UseMembersResult;
+
+function useMembers(
+  eventOrCallback?: MembersEvent | MembersEvent[] | UseMembersCallback,
+  callbackOrOptions?: UseMembersCallback | UseSpaceOptions,
+  optionsOrNothing?: UseSpaceOptions,
+): UseMembersResult {
+  const { space, connectionError, channelError } = useSpace();
+  const [members, setMembers] = useState<SpaceMember[]>([]);
+  const [others, setOthers] = useState<SpaceMember[]>([]);
+  const [self, setSelf] = useState<SpaceMember | null>(null);
+
+  const callback =
+    isString(eventOrCallback) || isArray(eventOrCallback) ? (callbackOrOptions as UseMembersCallback) : eventOrCallback;
+
+  const options = isFunction(callbackOrOptions) ? optionsOrNothing : callbackOrOptions;
+
+  const callbackRef = useRef<UseMembersCallback | undefined>(callback);
+
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
+
+  useEffect(() => {
+    if (callbackRef.current && members && !options?.skip) {
+      const listener: UseMembersCallback = (params) => {
+        callbackRef.current?.(params);
+      };
+      if (isString(eventOrCallback)) {
+        space?.members.subscribe(eventOrCallback, listener);
+      } else {
+        space?.members.subscribe<MembersEvent>(listener);
```

Comment:
`<MembersEvent>` should not be necessary if we use the type map https://github.com/ably/spaces/blob/main/src/Members.ts#L13-L19\

(well, at least I would hope so)

Reply:
I think it's impossible to calculate type right now. I would remove template type here I think it makes no sense, but I am not sure about backward compatibility

Reply:
could you clarify what "makes no sense` @ttypic ? The subscribe type is necessary because we want to use `.on/.off` while keeping the typing specified.

It allows you to have full type support when using subscribers:
<img width="894" alt="Screenshot 2023-10-05 at 17 22 35" src="https://github.com/ably/spaces/assets/1148118/d8b913bc-91cb-4059-ad87-d1d7c49b594f">

So when you type `update` you get the correct argument type.

`subscribe` itself is a compromise on our side, it's not pretty, but a simple `typeof` will not pass these event emitter types. And doing it as you did above is just a recipe for bugs as we have multiple sources of truth for the listeners and events.

Reply:
Why can it be just: 

```
subscribe(
    listenerOrEvents?: EventKey<MemberEventsMap> | EventKey<MemberEventsMap>[] | EventListener<SpaceMember>,
    listener?: EventListener<SpaceMember>,
  )
```

all events right now have exactly the same argument types. 

I don't know how TS can determine what is the actual type from

```
subscribe<K>(listener: EventListener<MemberEventsMap[K]>) 
```

imagine `MemberEventsMap` is something like:

```
type MemberEventsMap = {
  x1: X;
  x2: X;
 ...
  xN: X;
};
```

how to calculate `K`  if listener is `(param: X) => void` ? Is it `x1` or is it `x2` ...?

Reply:
probably it is better to use `any` or `// @ts-ignore`